vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/ScreenshotCropper.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.selenium.screenshot;
+
+import java.awt.Point;
+import java.awt.image.BufferedImage;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.openqa.selenium.WebDriver;
+import org.vividus.selenium.IWebDriverProvider;
+import org.vividus.ui.action.ISearchActions;
+import org.vividus.ui.action.search.Locator;
+
+import ru.yandex.qatools.ashot.coordinates.Coords;
+import ru.yandex.qatools.ashot.coordinates.CoordsProvider;
+
+public class ScreenshotCropper
+{
+    private final ScreenshotDebugger screenshotDebugger;
+    private final ISearchActions searchActions;
+    private final CoordsProvider coordsProvider;
+    private final IWebDriverProvider webDriverProvider;
+
+    public ScreenshotCropper(ScreenshotDebugger screenshotDebugger, ISearchActions searchActions,
+            CoordsProvider coordsProvider, IWebDriverProvider webDriverProvider)
+    {
+        this.screenshotDebugger = screenshotDebugger;
+        this.searchActions = searchActions;
+        this.coordsProvider = coordsProvider;
+        this.webDriverProvider = webDriverProvider;
+    }
+
+    public BufferedImage getScreenshot(BufferedImage image, Optional<Coords> contextCoords,
+            Map<IgnoreStrategy, Set<Locator>> ignoreStrategies, int topAdjustment)
+    {
+        BufferedImage outputImage = image;
+        WebDriver webDriver = webDriverProvider.get();
+
+        for (IgnoreStrategy strategy : List.of(IgnoreStrategy.ELEMENT, IgnoreStrategy.AREA))
+        {
+            Set<Coords> ignore = Optional.ofNullable(ignoreStrategies.get(strategy)).orElse(Set.of()).stream()
+                    .map(searchActions::findElements)
+                    .flatMap(Collection::stream)
+                    .collect(Collectors.collectingAndThen(
+                            Collectors.toList(),
+                            webElementsToIgnore -> coordsProvider.ofElements(webDriver, webElementsToIgnore)
+                    ));
+            if (!ignore.isEmpty())
+            {
+                Point adjustment = calculateAdjustment(contextCoords, topAdjustment);
+                ignore.forEach(c ->
+                {
+                    c.x += adjustment.x;
+                    c.y += adjustment.y;
+                });
+
+                outputImage = strategy.crop(outputImage, ignore);
+                screenshotDebugger.debug(this.getClass(), "cropped_by_" + strategy, outputImage);
+            }
```

Comment:
```suggestion
        WebDriver webDriver = webDriverProvider.get();
         Point adjustment = calculateAdjustment(contextCoords, topAdjustment);

        for (IgnoreStrategy strategy : List.of(IgnoreStrategy.ELEMENT, IgnoreStrategy.AREA))
        {
            Set<Coords> ignore = Optional.ofNullable(ignoreStrategies.get(strategy))
                                                                 .orElse(Set.of())
                                                                 .stream()
                                                                 .map(searchActions::findElements)
                                                                 .flatMap(Collection::stream)
                                                                 .map(e -> coordsProvider.ofElement(webDriver, e))
                                                                 .map(c -> {
                                                                    c.x += adjustment.x;
                                                                    c.y += adjustment.y;
                                                                     return c;
                                                                })
                                                                .collect(Collectors.toSet());
            if (!ignore.isEmpty())
            {
                outputImage = strategy.crop(outputImage, ignore);
                screenshotDebugger.debug(this.getClass(), "cropped_by_" + strategy, outputImage);
            }
```

Reply:
Because of what?

Reply:
just compare implementations, `coordsProvider.ofElementS` is not a simple loop